### PR TITLE
Switch internal handling of current deposition from compile-time to run-time

### PR DESCRIPTION
**Pull request by sbastrakov**
As we discussed on the VC, I split the transition into run-time current interpolation parameters into two PRs.

This PR only does the internal stuff and does *not* change the user interface of PIConGPU.
The only difference is that custom current deposition functors declared in `.param` files are no longer allowed (but I am not aware of such a use case at all).
It made internal changes necessary for a full transition of this parameter towards run-time, to be done in a following PR.

To avoid confusion, I renamed names and template parameter names of old instances of current interpolation into functors (which they were already and remain that way). Also due to a name conflict with .param file naming the new class is called `CurrentInterpolationInfo`, to be renamed into `CurrentInterpolation` in the follow-up PR.

### CODE REVIEW

File: include/picongpu/fields/currentInterpolation/CurrentInterpolation.hpp [hunk 1]
```diff
@@ -17,6 +17,77 @@
...
+            CurrentInterpolationInfo& operator=(CurrentInterpolationInfo const&) = delete;
+
+        private:
+            CurrentInterpolationInfo() = default;
+            ~CurrentInterpolationInfo() = default;
+        };
+
+    } // namespace currentInterpolation
+
+} // namespace picongpu
```
Comment (sbastrakov) [inline include/picongpu/fields/currentInterpolation/CurrentInterpolation.hpp:93]: Will fix

File: include/picongpu/fields/FieldJ.tpp [hunk 2]
```diff
@@ -250,8 +244,8 @@ namespace picongpu
         deposit.template execute<T_area, numWorkers>(cellDescription, depositionKernel, solver, jBox, pBox);
     }
 
-    template<uint32_t T_area, class T_CurrentInterpolation>
-    void FieldJ::addCurrentToEMF(T_CurrentInterpolation& myCurrentInterpolation)
+    template<uint32_t T_area, class T_CurrentInterpolationFunctor>
+    void FieldJ::addCurrentToEMF(T_CurrentInterpolationFunctor myCurrentInterpolationFunctor)
```
Comment (PrometheusPi) [inline include/picongpu/fields/FieldJ.tpp:248]: Why did you remove the by reference `&`?

**Review by PrometheusPi — APPROVED** [review]
Just a minor question.

File: include/picongpu/fields/FieldJ.tpp (hunk 2 above)
Comment (sbastrakov) [reply include/picongpu/fields/FieldJ.tpp:248]: It was a bit unnatural, but not incorrect originally. This parameter is a functor that's not only not modified internally, but is stateless generally. And with the changed version the call changed and it didn't compile with this reference to non-const. So changed to passing by value, whicj was imo better for the original version as well.

Comment (PrometheusPi) [reply include/picongpu/fields/FieldJ.tpp:248]: Thanks for the explanation.

**Review by psychocoderHPC — APPROVED** [review]
